packages/x-date-pickers/src/internals/models/value.ts

```diff
@@ -3,3 +3,17 @@ import { PickerValidDate } from '../../models/pickers';
 export type PickerValue = PickerValidDate | null;
 
 export type PickerRangeValue = [PickerValidDate | null, PickerValidDate | null];
+
+export type PickerNonNullableRangeValue = [PickerValidDate, PickerValidDate];
+
+export type InferPickerValue<TIsRange extends boolean> = TIsRange extends true
+  ? TIsRange extends false
+    ? PickerValue | PickerRangeValue
+    : PickerRangeValue
+  : PickerValue;
+
+export type InferNonNullablePickerValue<TIsRange extends boolean> = TIsRange extends true
```

Comment:
The fact that we need to put this logic and the `RangePosition` in the community package raises some "concern flags" in my head about the direction of the solution. 🙈